### PR TITLE
Implement view company command

### LINKED ISSUE #137: As a user, I want to view a company’s internships when I view their profile,
so that I can see what internships they are offering

### LINKED ISSUE #138: As a user, I want to view a company profile,
so that I can see its details

### PR DESCRIPTION
**Pull request by shawn-nyk**
PR overview:
* Implement view company command
* Make all view commands work
* Edit erroneous display messages

Resolves #137 
Resolves #138

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/view/ViewCompanyCommand.java
Comment (keanecjy) [inline src/main/java/seedu/address/logic/commands/view/ViewCompanyCommand.java]: Perhaps we can include a check for when the user is already currently viewing this current item and display a `MESSAGE_ALREADY_VIEWING` message?

**Review by keanecjy — COMMENTED** [review]
Great work! Just need @seanjyjy to check if it is alright before merging

**Comment by seanjyjy** [comment]
Besides the point that @keanecjy has raised, I believe it is good to merge after it is resolved.

File: src/main/java/seedu/address/logic/commands/view/ViewCompanyCommand.java
Comment (shawn-nyk) [reply src/main/java/seedu/address/logic/commands/view/ViewCompanyCommand.java]: Oh yep great idea! I've made the changes!

**Comment by shawn-nyk** [comment]
> Besides the point that @keanecjy has raised, I believe it is good to merge after it is resolved.

Thanks Sean! Have made the changes!

File: src/main/java/seedu/address/logic/commands/view/ViewApplicationCommand.java
Comment (seanjyjy) [inline src/main/java/seedu/address/logic/commands/view/ViewApplicationCommand.java]: just a minor question but is there a need for this if block? I think we could just shift it into the if-else block. However, if this is a better coding standard, then I think we can stick to this.

Comment (shawn-nyk) [reply src/main/java/seedu/address/logic/commands/view/ViewApplicationCommand.java]: I can't shift it straight into the current if-else block due to the nature of the condition checking. If we want to make it all part of one block, here's an option:
```
if (model.getTabName() != TabName.APPLICATION) {
    model.setTabName(TabName.APPLICATION);
    shouldSwitchTab = true;
    model.setViewIndex(targetIndex);
} else if (model.getViewIndex().equals(targetIndex)) {
    resultMessage = messageAlreadyViewing;
    shouldSwitchDisplay = false;
} else {
    model.setViewIndex(targetIndex);
}
```
But this means `model.setViewIndex(targetIndex)` appears twice.

Another option is to simply remove the if condition and do:
```
if (model.getTabName() != TabName.APPLICATION) {
    model.setTabName(TabName.APPLICATION);
    shouldSwitchTab = true;
    model.setViewIndex(targetIndex);
} else if (model.getViewIndex().equals(targetIndex)) {
    resultMessage = messageAlreadyViewing;
    shouldSwitchDisplay = false;
}

model.setViewIndex(targetIndex);
```
But this means that even when the model's view index is already the index that the user inputs, we still set it anyway (unnecessary call to setViewIndex). It won't affect the program though.

Between the current code and these 2 options, which do yall think is better? Or are there other alternatives yall may have?

Comment (seanjyjy) [reply src/main/java/seedu/address/logic/commands/view/ViewApplicationCommand.java]: oh if this the case, I believe the current code is a better option

File: src/main/java/seedu/address/logic/commands/view/ViewCompanyCommand.java
Comment (keanecjy) [inline src/main/java/seedu/address/logic/commands/view/ViewCompanyCommand.java]: I think that it might be better if we initialize the variables in the if-else block, so that it is clearer which block will have what booleans. And also this will stick to the notion of defining the variables in the least possible scope!

File: src/main/java/seedu/address/logic/commands/view/ViewApplicationCommand.java
Comment (keanecjy) [reply src/main/java/seedu/address/logic/commands/view/ViewApplicationCommand.java]: I think the current code is better. Also, why is there checks for tabName in this command? Since all the other commands automatically switches to the correct tab, even if it is already on the right tab

File: src/main/java/seedu/address/logic/commands/view/ViewCompanyCommand.java
Comment (shawn-nyk) [reply src/main/java/seedu/address/logic/commands/view/ViewCompanyCommand.java]: Hi Keane! Unfortunately that would not be possible. Variables can't be declared in an if-else block and then used outside of it due to variable scope. Even doing something as simple as:
```
if (true) {
    boolean shouldSwitchTab = false;
} else {
    boolean shouldSwitchTab = false;
}
shouldSwitchTab = true; // line X
```
Will lead to an error on line X whereby it doesn't detect that shouldSwitchTab has been initialized.
(More on it here: [https://stackoverflow.com/questions/17444020/why-cant-variables-be-declared-in-an-if-statement](https://stackoverflow.com/questions/17444020/why-cant-variables-be-declared-in-an-if-statement#:~:text=Because%20when%20b%20goes%20out,in%20the%20next%20if%20statement.&text=So%20here%20scope%20of%20b,it%20is%20giving%20compilation%20error.))

As such, this already is the smallest possible scope!

Comment (shawn-nyk) [reply src/main/java/seedu/address/logic/commands/view/ViewCompanyCommand.java]: To justify having these boolean vars initialized and declared where they are, we can take it that by default, we assume the user in on the appropriate tab (`shouldSwitchTab = false`) and not viewing the current item (`shouldSwitchDisplay = true`) which is why they would want to execute the view command in the first place. However, they may happen to be on the wrong tab (first if statement checks for that), or they may happen to be on the right tab and already viewing the item (the following else-if statement checks for that). That is how I would intend the code to be read!

File: src/main/java/seedu/address/logic/commands/view/ViewApplicationCommand.java
Comment (shawn-nyk) [reply src/main/java/seedu/address/logic/commands/view/ViewApplicationCommand.java]: Ah good question! I raised this in our Tele group last night, but basically how the view index works is that it takes the index from the tab that is currently in view. So for example, if a user is on say the Application tab and he is viewing the 3rd Application, if he executes `view com 3`, if we do not check for the tab first, then the program will simply think we're already viewing the 3rd item so it doesn't need to do anything, which is wrong of course. That said, now that you've raised this point, I believe I may be able to abstract away the first if statement, but I'll definitely need that condition check in the else-if statement that checks for if the user is on the same tab and currently viewing the indexed item. So I can minimize the tab checking but we need that one minimally.

File: src/main/java/seedu/address/logic/commands/view/ViewCompanyCommand.java
Comment (keanecjy) [reply src/main/java/seedu/address/logic/commands/view/ViewCompanyCommand.java]: Oh what I meant was initializing the value in the if-else block and not declaring it there. So an example of this is:

image

File: src/main/java/seedu/address/logic/commands/view/ViewApplicationCommand.java
Comment (keanecjy) [reply src/main/java/seedu/address/logic/commands/view/ViewApplicationCommand.java]: Oh okay thanks for the clarification!

File: src/main/java/seedu/address/logic/commands/view/ViewCompanyCommand.java
Comment (shawn-nyk) [reply src/main/java/seedu/address/logic/commands/view/ViewCompanyCommand.java]: Ah my bad! But then will we be violating this coding standard rule?
image

File: src/main/java/seedu/address/logic/commands/view/ViewApplicationCommand.java
Comment (shawn-nyk) [reply src/main/java/seedu/address/logic/commands/view/ViewApplicationCommand.java]: No problem, thanks for highlighting! Have made the changes!

File: src/main/java/seedu/address/logic/commands/view/ViewCompanyCommand.java
Comment (keanecjy) [reply src/main/java/seedu/address/logic/commands/view/ViewCompanyCommand.java]: Don't think so, since he did mention that this is just a guide and I remembered there was an example which he said would be ok?

**Review by keanecjy — APPROVED** [review]
LGTM

File: src/main/java/seedu/address/logic/commands/view/ViewCompanyCommand.java
Comment (shawn-nyk) [reply src/main/java/seedu/address/logic/commands/view/ViewCompanyCommand.java]: Oh if that's the case, sure I don't mind changing it! But then again I'm not sure if our evaluators have the same understanding though? For example when we eventually evaluate our peers' work, I would probably follow the coding standard to cross-check my evaluation and if I spotted this violation in their work, I would probably just mark it as a violation since it is stated in black and white and not have to think too much. If you have the source, could you point me to where Prof Damith mentioned that we could do otherwise? Would be much appreciated!

**Pull request merged**
